### PR TITLE
feat(tracing): node-postgres (pg) instrumentation

### PR DESCRIPTION
**Pull request by pglombardo**
To Do:

- [x] More query types: string, config object
- [x] Add instrumentation/tests for Promise based queries
- [x] Validate queries using both `Pool` & `Client` separately
- [x] Validate tracing of stored procedure calls
- [x] Validate transactional use case
- [x] Validate error capture/reporting
- [x] Validate query parsing/filtering
- [x] Validate vanilla tracing (when not tracing)
- [x] Update documentation

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -43,3 +43,14 @@ services:
...
+  postgres:
+    image: postgres:10.4
+    ports:
+      - 5432:5432
+    environment:
+      POSTGRES_USER: node
+      POSTGRES_PASSWORD: nodepw
+      POSTGRES_DB: nodedb
+    volumes:
+      - /srv/docker/postgresql:/var/lib/postgresql
```
Comment (basti1302) [inline docker-compose.yml]: What's that `volume` for? This fails for me locally. I guess it can be removed, right?

Comment (pglombardo) [reply docker-compose.yml]: That should give persistent data.  I'll check if we really need it.

**Comment by pglombardo** [comment]
This is ready for review/merge and release hopefully tomorrow.

File: src/tracing/instrumentation/pg.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  instrumentClient(pg.Client);
+}
+
+function instrumentClient(Client) {
+  shimmer.wrap(Client.prototype, 'query', shimQuery);
+}
+
+function shimQuery(original) {
+  return function() {
+    logger.warn('shimQuery: isTracing() == ' + cls.isTracing());
```
[the remark this answers is not in the document]
Comment (pglombardo) [reply src/tracing/instrumentation/pg.js]: doh I need to remember to test without expecting debug output.  I'll remove.

File: src/tracing/instrumentation/pg.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  }
+
+  var host = ctx.connectionParameters.host;
+  var port = ctx.connectionParameters.port;
+  var user = ctx.connectionParameters.user;
+  var db = ctx.connectionParameters.database;
+
+  return cls.ns.runAndReturn(function() {
+    var span = cls.startSpan('postgres');
+    span.b = {s: 1};
```
[the remark this answers is not in the document]
Comment (pglombardo) [reply src/tracing/instrumentation/pg.js]: For Python, Ruby and Go I never add batch unless there is actual batching so I would say no.  Plus adding batch KVs would be a waste of storage space with billions of spans - "not-batched" should be assumed if it's not specified.

**Comment by pglombardo** [comment]
Thanks for the feedback @basti1302 !  Updated without debug output log lines.

**Comment by pglombardo** [comment]
To be merged post release: https://github.com/instana/docs/pull/196

File: src/tracing/instrumentation/pg.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (pglombardo) [reply src/tracing/instrumentation/pg.js]: Good catch @basti1302  - fixed.

**Comment by basti1302** [comment]
This has landed in 1.41.0

**Pull request merged**